WIP: Codegen: clean up and remove old files

thockin opened the pull request:
I noticed a bunch of files that do not get re-generated.  I asked @liggitt @lavalamp and @jpbetz about it and it seems some are old and should be removed.

I decided to try to make this automatic.  Step 1 was to remove everything that appears auto-generated and run each generator to see what comes back.  In doing this I found a few bugs and other less-than-ideal cases.

WIP: This PR is currently just the remove/re-add.  I will automate it, but I want to a) see what barfs in CI and b) solicit feedback on it.  Specifically:

* update-mocks.sh doesn't actually know which files were generated, so it looks at git diff, which misses newly-added files, which then do not get the boilerplate.  I fixed it, kind of, but it's still not pretty.
* update-generated-kubelet-plugin-registration and update-generated-pod-resources both were missing directives to generate properly (including a v1 API !!).  These scripts are far too fragile.
* One example was missing a generate directive

Overall it is (IMO) far too hard to know everything that is auto-generated.  We do not have consistent naming.  Does anyone object if I try to do better here?

/kind bug
/kind cleanup

```release-note
NONE
```

TODO:
* Scripts which grep for "DO NOT EDIT" causes verify-boilerplate to flag them (script files) as "generated", becuase it greps (indiscriminately) for "DO NOT EDIT" itself (and so it matches the "grep" command in the scripts).
* The subproject verify scripts need to be touched or dumped
* Maybe dump all the subproject stuff
* The mocks and stable-metrics need to be touched
* All the currently deleted "deprecated" trees are suspect.
* Strip out the "remove all" and "regen X" commits and just leave the meaningful changes
* Use `:(glob)` in git-ls-files for max correct behavior

liggitt commented on hack/update-codegen.sh:
```diff
@@ -42,65 +42,34 @@ if [[ "${DBG_CODEGEN}" == 1 ]]; then
+    # Similar to find but faster and easier to understand.  We want to include
+    # modified and untracked files because this might be running against code
+    # which is not tracked by git yet.
+    git ls-files -cmo --exclude-standard "$@"
```
this is magical...

liggitt commented on hack/update-codegen.sh:
```diff
@@ -693,6 +678,10 @@ function codegen::listers() {
     fi
 
+    git_find -z 'staging/src/k8s.io/client-go/**/*.go' \
+        | { xargs -0 grep -z -l "^// Code generated by lister-gen. DO NOT EDIT.$" || true; } \
```
can we scope to `k8s.io/client-go/listers` and delete all go files?

liggitt commented on hack/update-codegen.sh:
```diff
@@ -729,6 +716,10 @@ function codegen::informers() {
     fi
 
+    git_find -z 'staging/src/k8s.io/client-go/**/*.go' \
+        | { xargs -0 grep -z -l "^// Code generated by informer-gen. DO NOT EDIT.$" || true; } \
```
scope to `k8s.io/client-go/informers` and delete all go files?

liggitt commented on hack/update-generated-kubelet-plugin-registration-dockerized.sh:
```diff
@@ -23,13 +23,17 @@ set -o nounset
-KUBELET_EXAMPLE_PLUGIN_V1BETA2="${KUBE_ROOT}/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta2/"
+
+dirs=(
+    "staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1alpha1/"
+    "staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1beta1/"
+    "staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1/"
```
is it possible to make this list directories under `staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v*` or something so we pick up all versioned subdirs, including any future ones? same question for example_plugin_apis and podresources below

thockin replied:
yes, I want to followup and not make this a hardcoded list.

Also, we have like 5 different scripts which all run the same command on different dirs - maybe just one script (update-codegen) ?

liggitt commented on hack/update-generated-protobuf.sh:
```diff
@@ -27,12 +27,21 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+    git ls-files -cmo --exclude-standard ':!:vendor/**' "$@"
+}
+
+git_find -z '**/generated.proto' | xargs -0 rm -f
+git_find -z '**/generated.pb.go' | xargs -0 rm -f
```
any way to scope this to APIROOTS packages?

liggitt commented on hack/update-mocks.sh:
```diff
@@ -84,14 +100,14 @@ for IFILE in $(find_files | xargs grep --files-with-matches -e '//go:generate mo
-  fi
+  # only process files that appear to be mocks
+  test -f "$file" || continue
+  grep -q "^// Code generated by MockGen. DO NOT EDIT.$" "$file" || continue
```
would limiting this to .go files and fuzzy matching the comment (e.g. "^//.*generated by MockGen") be less fragile?

thockin replied:
Note: this is only evaluating modifed or new (not tracked) files, which SHOULD only be the newly generate mocks.  Not sure extra fuzziness helps - that is the string being used and if it changes I think we WANT to know..  At least one file looks like it was generated by the tool and committed without a regen possible.

liggitt replied:
ok... will anything notice if the mockgen comment changes and this script doesn't change?

thockin replied:
Assuming we bumped the depency to a version thast used a different string, CI would fail verify-mocks.sh, so we'd at least have to go and regenerate all the mockfiles, presumably with this script.  Since the comment would have changed, this script will not find them and will therefore not add boilerplate.  So even if someone bypassed all that and then commited them, it would fail boilerplate.

So ... "yes, I think so" :)

thockin commented on hack/update-codegen.sh:
```diff
@@ -611,6 +588,10 @@ function codegen::applyconfigs() {
     fi
 
+    git_find -z 'staging/src/k8s.io/client-go/**/*.go' \
+        | { xargs -0 grep -z -l "^// Code generated by applyconfiguration-gen. DO NOT EDIT.$" || true; } \
```
I agree that grepping for comments is sort of awful, but I don't want to start open-coding exclusions.  These things should all be as metadata-driven as possible.

Here's an alternative: What if we renamed all of these output files?  Instead of `staging/src/k8s.io/client-go/applyconfigurations/apps/v1/deployment.go` maybe `staging/src/k8s.io/client-go/applyconfigurations/apps/v1/generated.deployment.go` or `deployment.generated.go` or `zz_generated.applyconfig.deployment.go` or something?  How much would people hate that?

thockin replied:
Indeed, grepping for "DO NOT EDIT" causes verify-boilerplate to flag these files as "generated", becuase it greps (indiscriminately) for "DO NOT EDIT" (so it matches the "grep" command in the scripts).  BLeeeeeech.

thockin replied:
2 ideas here:

1) We make sure EVERY generated file has a discoverable name.  I don't know what this would do for things like proto files which are generated (isn't filename important in proto?) but it would work for .go files.

2) We make sure EVERY generated file has a CLEARLY documented comment that is not easy to mistake.  E.g. `^[/*# ]+ THIS FILE IS AUTO-GENERATED: DO NOT EDIT$` and maybe another line like  `^[/*# ]+ +k8s:generated-by: (.*)$` where the match-group is the tool name for easier grepping.  Basically, write a spec for genfile detection.

3) rot13 the strings in shell scripts :)

liggitt replied:
Either a fixed filename or a fixed generated token in the file (like the `+k8s:generated-by: ...` bit) we can match on seem reasonable. I'd probably support both approaches so we don't have to rename the world and can accommodate generated files where the filename is important/fixed for some other reason.

thockin replied:
Either approach is pretty enormous.  To stick such a tag in emitted files means touching gengo and we still have the problem that the script are grepping for a string which other tools will find in those scripts.

I'm going to see if there's a less dramatic answer if we just hold our noses a bit.  Really, the boilerplate script should just ignore diffs that are "THIS FILE IS AUTO-GENERATED".

thockin replied:
I think I can make boilerplate.py smarter.

liggitt replied:
> we still have the problem that the script are grepping for a string which other tools will find in those scripts

excluding non-go files or the hack dir seems like it would generally work, no?

> I think I can make boilerplate.py smarter.

that works as well

thockin commented on hack/update-codegen.sh:
```diff
@@ -657,6 +638,10 @@ function codegen::clients() {
         done
     fi
 
+    git_find -z 'staging/src/k8s.io/client-go/**/*.go' \
```
That would be cleaner if we could follow that pattern but 2 odd cases:

1) applyconfig has a manual top-level doc.go - maybe they all should (applyconfig, listers, informers, clients) instead of generating it?  Then it's not a one-off exclusion any more.

Some of the subprojects have "deprecated" trees that are not regenerated - I can't tell if those should be removed or un-tagged as "generated by"?  E.g. staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/deprecated/...

liggitt replied:
> applyconfig has a manual top-level doc.go - maybe they all should (applyconfig, listers, informers, clients) instead of generating it? Then it's not a one-off exclusion any more.

+1 to consistent doc.go file to make the package resolvable to go tools even when all the generated files are removed

> Some of the subprojects have "deprecated" trees that are not regenerated - I can't tell if those should be removed or un-tagged as "generated by"? E.g. staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/deprecated/...

probably remove them... I think they got missed in #95923

thockin replied:
Will remove deprecated.

thockin committed:
Retool subproject client generation

thockin commented:
Just repushed.

I don't think I addressed any of your comments yet, but I merged all 5 "run protoc" update/verify into one script and call it from update-codegen now, so muuuuuuch simpler.

Not done - the subproject verify scripts need to be touched, mocks and stable-metrics need to be touched, and all the currently deleted "deprecated" trees are suspect.

When all that is done I can strip out the "remove all" and "regen X" commits and just leave the meaningful changes.

Happy to take feedback in the meantime.

cici37 commented:
/triage accepted

liggitt commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.pb.go:
```diff
@@ -9653,7 +9653,7 @@ func init() {
 
 var fileDescriptor_00212fb1f9d3bf1c = []byte{
-	// 6539 bytes of a gzipped FileDescriptorProto
+	// 6514 bytes of a gzipped FileDescriptorProto
```
*raises eyebrow*

was this stale?

thockin replied:
No.  Another rabbithole.  It's encoded gzipped data.  Apparently gzip is not deterministic (so makes a GREAT choice for this embedded data /s) and this blob encodes the go_package.  If you decode and diff this blob the delta is:

```diff
@@ -4684,7 +4684,7 @@
         }
       ],
       "options": {
    -    "go_package": "k8s.io/cri-api/pkg/apis/runtime/v1"
    +    "go_package": "v1"
       },
       "syntax": "proto3"
     }
 ```

I put this in the commit message for "Move update-generated-runtime into common script".

What I can't tell is if this really matters.

liggitt replied:
`"go_package": "v1"` looks wrong... isn't that supposed to be the full package?

thockin replied:
Yeah - I don't know how this is used.  I am not sure on it and need to figure it out. 

I ran the same decode on ALL the pb.go files and only one other ("dra") sets this option, and it sets it to "dra".

PB code documents this field as:

```
    // Sets the Go package where structs generated from this .proto will be
    // placed. If omitted, the Go package will be derived from the following:
    //   - The basename of the package import path, if provided.                                                
    //   - Otherwise, the package statement in the .proto file, if present.
    //   - Otherwise, the basename of the .proto file, without extension.
```

Which, to my read, sounds like "v1" is correct and "k8s.io/cri-api/pkg/apis/runtime/v1" was always wrong.

Your opinion?

liggitt replied:
but... it doesn't match the proto file, right?

https://github.com/kubernetes/kubernetes/blob/a66aad2d80dacc70025f95a8f97d2549ebd3208c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L21

https://github.com/kubernetes/kubernetes/pull/106630 pointed at https://developers.google.com/protocol-buffers/docs/reference/go-generated#package which said:

> The Go import path is locally specified in a .proto file by declaring a go_package option with the full import path of the Go package. Example usage:
>
> `option go_package = "example.com/project/protos/fizz";`

thockin replied:
Indeed.  I have reached out to Go libs team (but they are in EU,  so it's weekend).  It does seem like a full path is more common.

Several of these kubelet gRPC APIs don't specify this option at all, and it's not clear to me what impact that has.

The problem is that along the way the directory and the package get concatenated  so we end up with staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/k8s.io/cri-api/pkg/apis/runtime/v1/api.pb.go

thockin replied:
I can reformulate how protoc is called and remove this diff.  Will be in my next push

thockin commented on hack/lib/golang.sh:
```diff
@@ -521,7 +521,10 @@ kube::golang::setup_env() {
-  export GOCACHE="${KUBE_GOPATH}/cache"
+  # If these are not set, set them now.  THis ensures that any subsequent
+  # scripts we call (which may call this function again) use the same values.
+  export GOCACHE="${GOCACHE:-"${KUBE_GOPATH}/cache/build"}"
+  export GOMODCACHE="${GOMODCACHE:-"${KUBE_GOPATH}/cache/mod"}"
```
Yeah, this is kind of a mess because of a lot of reasons.  Until workspaces lands, we are forced to phony-up a GOPATH.  Go's default is to put the module cache under GOPATH[0], which means it gets stored in this directory tree.

Running in a work tree and then running one of these scripts means the GOPATH changes and the modcache is found to be empty, so it re-downloads.

This change explicitly points the caches to the "real" tree, so we avoid the re-download, but as you point out, it is not very well managed.

I wish I could say workspaces fixes this, but it doesn't.  That tree (for now) STILL sets up a phony GOPATH, though for a different reason - `go install`.  We can't just set GOBIN (not compatible with crossbuilds).  We *could* use `go build -o` but then EVERYONE needs to get it just right (and a trailing slash matters, if I recall).  So the only way I found to make it easy to build tools was to set GOPATH and use `go install`, which writes to GOPATH[0]/bin.  PUUUUUUUKE.  I'd much rather run with GOPATH=/dev/null.  Better ideas are welcome :)

thockin commented:
Once this is all happy I will remove TMP commits and break into a few smaller PRs (not much smaller, sorry)

thockin committed:
Update vendor

thockin commented:
85% confident this one passes

thockin committed:
Update publishing bot deps

thockin commented:
Down to a sort-order issue.  95% confidence this time.

thockin commented:
The sort issue is on my end, sigh

thockin commented:
Rebased, cleaned up, sort should be stable now.  Let's see.

thockin commented:
w00t!  All green

nikhita commented on staging/publishing/rules.yaml:
```diff
@@ -397,8 +397,6 @@ rules:
       branch: master
     - repository: kms
       branch: master
-    - repository: code-generator
-      branch: master
```
code-generator needs to be listed as a dependency due to https://github.com/kubernetes/kubernetes/commit/321095dccb60a14e4a886296014ff014759f52b9

liggitt replied:
I think this PR drops the code-generator-using scripts from those repos, so they no longer have content that references code-generator, so this looks right

thockin replied:
correct - I intend to retool this all in go workspaces conversion, to erase any dependencies on GOPATH or `../../../../` sorts of trickery

liggitt commented on hack/update-codegen.sh:
```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
+        return 0
+    fi
+
+    GO111MODULE=on GOPROXY=off \
```
GOPROXY=off is surprising... what's up with that (same question applies elsewhere)?

thockin replied:
That's a pattern copied from the original source.  These probably ALL should move from `go install` to `hack/make-rules/build.sh`.  Can be a pre-factoring.

thockin replied:
https://github.com/kubernetes/kubernetes/pull/115171

liggitt commented on hack/update-codegen.sh:
```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
     fi
 }
 
+# $1: output package base (in Go pkg syntax)
+# $2: input package base (in Go pkg syntax)
```
are these positional ($1 and $2) or flags (`--output_pkg_root` and `--input_pkg_root`)?

thockin replied:
forgot to remove those - from previous drafts

liggitt commented on hack/update-codegen.sh:
```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
+#   --input_pkg_root <string>
+#     The root directory (below staging/src) from which to search for types.go
+#     files which request clients to be generated.  This must be Go package
+#     sytax, e.g. "k8s.io/foo/bar".
```
```suggestion
#     syntax, e.g. "k8s.io/foo/bar".
```

thockin replied:
ACK

liggitt commented:
> Assuming it passes - @liggitt - how many PRs do you want? I could break this up into as many as 30 PRs or as few as 4 or 5.

Ummmm..... fewer PRs with independent commits that are trivially reviewable or non-overlapping seems good, with independent cleanup stuff that doesn't mess with generated files hoisted into prereq PRs?

thockin commented on hack/update-codegen.sh:
```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
+#
+#   --output_pkg_root <string>
+#     The root directory (below staging/src) into which generated directories
+#     and files will be placed.  This must be Go package sytax, e.g.
```
ACK

thockin commented on hack/update-codegen.sh:
```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
+# $2: input package base (in Go pkg syntax)
+# Generate client code
+#
+# Args:
```
ACK

thockin commented:
I'll start breaking it up later today or tomorrow, will ping here.  Meantime, any furhter comments you have on this commits will get folded in.

bart0sh commented:
/priority important-longterm

thockin commented:
Now that most of this has been merged in other PRs, I will revisit the subproject codegen.  It's ugggggggly but it seems to work AS LONG AS you are in a GOPATH.


I am thinking:

1) keep it working (so you can clone kube-aggregator and run a local update-codegen script)
2) convert those verify-codegen scripts to use worktree (so they are trivial and mostly identical)
3) clean up those update-codegen scripts as much as I can

thockin commented:
How committed are we to this pattern where an end-user repo calls k8s.io/code-generator/generate-internal-groups.sh ?  Do we consider that an API we guarantee compat around, or are we willing to change it?

I ask because it's fairly convoluted and hard to navigate, and I feel like it COULD be simpler, but I don't want to waste (more) time on a dead end.  

Also, as part of go workspaces, I think the deepcopy etc tools need to change their CLIs anyway.

liggitt commented:
> How committed are we to this pattern where an end-user repo calls k8s.io/code-generator/generate-internal-groups.sh ? Do we consider that an API we guarantee compat around, or are we willing to change it?

I suspect any usable alternative would be fine, but would defer to @sttts

thockin committed:
Codegen: generate-internal-groups is deprecated

This folds everything back into generate-groups.sh.  Callers who used
"all" have been converted to list the generators - it's safer unless we
want to add the extra generated files for them.

thockin committed:
Codegen: subprojects: nuke existing files

thockin committed:
Neutralize "internal" in generate-internal-groups

This kills most of the deltas between generate-groups.sh and
generate-internal-groups.sh.  That will be cleaned up more in later
commits.

Since this constitutes an external API, the calling args are retained.

thockin committed:
------- TODO FILE -----

thockin committed:
Codegen: subproject: chdir to the project dir

This makes it feel more like an API to be used in other repos (which is
really what it is for).

thockin committed:
Codegen: clean up stray ls-files use

thockin committed:
Codegen: deepcopy needs to run before swagger

thockin committed:
Codegen: git grep returns non-zero if no match

thockin committed:
Codegen: easier to read logs for subproject output

thockin committed:
Codegen: subprojects: clean up verify scritps

They all run successfully.

thockin committed:
Codegen: simpler k8s.io/code-generator/examples

This moves the hack/ directory and scripts to the examples dir, which is
a distinct module.  This avoids some Go unpleasantness around module
boundaries and just makes more sense.
